File: test/idManager.test.js
import assert from "assert";
import { randomBytes } from "crypto";
import IdManager from "../src/IdManager";
import VaultysId from "../src/VaultysId";
import { MemoryChannel } from "../src/MemoryChannel";
import { MemoryStorage }  from "../src/MemoryStorage";

describe("IdManager", () => {
  it("serder a vaultys secret", async () => {
    const id1 = await VaultysId.generateMachine();
    const secret = id1.getSecret();
    const id2 = VaultysId.fromSecret(secret);
    assert.equal(id2.fingerprint, id1.fingerprint);
    assert.equal(id2.id.toString("base64"), id1.id.toString("base64"));
    assert.deepStrictEqual(id2.didDocument, id1.didDocument);
  });

  it("serder a vaultys secret in base64", async () => {
    const id1 = await VaultysId.generateMachine();
    const secret = id1.getSecret("base64");
    const id2 = VaultysId.fromSecret(secret, "base64");
    assert.equal(id2.fingerprint, id1.fingerprint);
    assert.equal(id2.id.toString("base64"), id1.id.toString("base64"));
    assert.deepStrictEqual(id2.didDocument, id1.didDocument);
  });

  it("serder to public Idmanager", async () => {
    const id1 = await VaultysId.generateMachine();
    const id2 = VaultysId.fromId(id1.id);
    assert.equal(id2.fingerprint, id1.fingerprint);
    assert.equal(id2.id.toString("base64"), id1.id.toString("base64"));
    assert.deepStrictEqual(id2.didDocument, id1.didDocument);
  });

  it("serder to public Idmanager stringified", async () => {
    const id1 = await VaultysId.generateMachine();
    const id = JSON.stringify(id1.id);
    const id2 = VaultysId.fromId(JSON.parse(id));
    assert.equal(id2.fingerprint, id1.fingerprint);
    assert.equal(id2.id.toString("base64"), id1.id.toString("base64"));
    assert.deepStrictEqual(id2.didDocument, id1.didDocument);
  });

  it("serder to public Idmanager as hex string", async () => {
    const id1 = await VaultysId.generateMachine();
    const id2 = VaultysId.fromId(id1.id.toString('hex'));
    assert.equal(id2.fingerprint, id1.fingerprint);
  });

  it("serder to public Idmanager as base64 string", async () => {
    const id1 = await VaultysId.generateMachine();
    const id2 = VaultysId.fromId(id1.id.toString('base64'), null, "base64");
    assert.equal(id2.fingerprint, id1.fingerprint);
    assert.equal(id2.id.toString("base64"), id1.id.toString("base64"));
    assert.deepStrictEqual(id2.didDocument, id1.didDocument);
  });

  it("sign unspecified data and log it in the store", async () => {
    const s = MemoryStorage(() => "");
    const manager = new IdManager(await VaultysId.generatePerson(), s);
    const signature = await manager.signChallenge(manager.vaultysId.id);
    const signatures = manager.getSignatures();
    assert.equal(signatures.length, 1);
    assert.equal(signatures[0].payload.challenge.toString("hex"), manager.vaultysId.id.toString("hex"));
    assert.equal(signatures[0].payload.signature.toString("hex"), signature.toString("hex"));
    assert.equal(signatures[0].type, 'UNKNOWN');
  });

  it("sign random document hash and log it in the store", async () => {
    const s = MemoryStorage(() => "");
    const fileHashMock = randomBytes(32);
    const manager = new IdManager(await VaultysId.generatePerson(), s);
    const payload = await manager.signFile(fileHashMock);
    const signatures = manager.getSignatures();
    assert.equal(signatures.length, 1);
    const challenge = new URL(signatures[0].challenge)
    assert.equal(challenge.searchParams.get('hash'), fileHashMock.toString("hex"));
    assert.ok(manager.verifyFile(signatures[0].payload.challenge.toString("hex"), signatures[0].payload.signature));
    assert.equal(signatures[0].payload.signature.toString("hex"), payload.signature.toString("hex"));
    //console.log(signatures[0]);
    assert.equal(signatures[0].type, 'DOCUMENT');
  });

  // it("sign login and log it in the store", async () => {
  //   const s = MemoryStorage(() => "");
  //   const loginMock = 'vaultys://login?host=https://sso.vaultys.net/interaction/1UfGEF9HDQiIreFPS3wlI&nonce=9c0c7621a790c6e697032093aeca614d183319d663aa5cc1a085e052c7f904d5&timestamp=1665498137687&challenge=dummy'
  //   const manager = new IdManager(await VaultysId.generatePerson(), s);
  //   const payload = await manager.signLogin(loginMock);
  //   const signatures = manager.getSignatures();
  //   assert.equal(signatures.length, 1);
  //   const challenge = new URL(signatures[0].challenge)
  //   assert.equal(challenge.searchParams.get('hash'), fileHashMock.toString("hex"));
  //   assert.ok(manager.verifyFile(signatures[0].payload.challenge.toString("hex"), signatures[0].payload.signature));
  //   assert.equal(signatures[0].payload.signature.toString("hex"), payload.signature.toString("hex"));
  //   //console.log(signatures[0]);
  //   assert.equal(signatures[0].type, 'LOGIN');
  // });
});

describe("SRG challenge with IdManager", () => {
  it("pass a challenge", async () => {
    const channel = MemoryChannel.createBidirectionnal();
    const s1 = MemoryStorage(() => "");
    const s2 = MemoryStorage(() => "");
    const manager1 = new IdManager(await VaultysId.generatePerson(), s1);
    const manager2 = new IdManager(await VaultysId.generateMachine(), s2);
    const metadata1 = {
      name: "a",
      email: "b",
      phone: "c"
    };
    const metadata2 = {
      name: "d",
      email: "e",
      phone: "f"
    };

    const contacts = await Promise.all([
      manager1.askContact(channel, metadata1),
      manager2.acceptContact(channel.otherend, metadata2),
    ]);

    assert.equal(contacts[0].did, manager2.vaultysId.did);
    assert.equal(contacts[1].did, manager1.vaultysId.did);

    // console.log(s2.substore("contacts"))
    
    // assert.deepStrictEqual(s2.substore("contacts").get(manager1.vaultysId.did).metadata, metadata1);
    // assert.deepStrictEqual(s1.substore("contacts").get(manager2.vaultysId.did).metadata, metadata2);

    assert.equal(Object.values(s1.substore("wot")._raw).length, 1);
    assert.equal(Object.values(s2.substore("wot")._raw).length, 1);
    
    assert.equal(manager2.contacts.length, 1);
    assert.equal(manager2.getContact( manager1.vaultysId.did).fingerprint, manager1.vaultysId.fingerprint);
    
    manager2.setContactMetadata(manager1.vaultysId.did, "name", "salut");
    manager2.setContactMetadata(manager1.vaultysId.did, "group", "pro");
    assert.equal(manager2.getContactMetadata(manager1.vaultysId.did, "name"), "salut");
    assert.equal(manager2.getContactMetadata(manager1.vaultysId.did, "group"), "pro");

    assert.ok(await manager1.verifyRelationshipCertificate(manager2.vaultysId.did));
    assert.ok(await manager2.verifyRelationshipCertificate(manager1.vaultysId.did));

  });

  it("pass a challenge over encrypted Channel", async () => {
    const channel = MemoryChannel.createBidirectionnal();

    const s1 = MemoryStorage(() => "");
    const s2 = MemoryStorage(() => "");
    const manager1 = new IdManager(await VaultysId.generatePerson(), s1);
    const manager2 = new IdManager(await VaultysId.generateOrganization(), s2);
    const metadata1 = {
      name: "a",
      email: "b"
    };
    const metadata2 = {
      name: "d",
      phone: "f"
    };

    const contacts = await Promise.all([
      manager1.askContact(channel, metadata1),
      manager2.acceptContact(channel.otherend, metadata2),
    ]);

    assert.equal(contacts[0].did, manager2.vaultysId.did);
    assert.equal(contacts[1].did, manager1.vaultysId.did);

    // assert.deepStrictEqual(s2.substore("contacts").get(manager1.vaultysId.did).metadata, metadata1);
    // assert.deepStrictEqual(s1.substore("contacts").get(manager2.vaultysId.did).metadata, metadata2);

    assert.equal(Object.values(s1.substore("wot")._raw).length, 1);
    assert.equal(Object.values(s2.substore("wot")._raw).length, 1);

    manager1.setContactMetadata(manager2.vaultysId.did, "name", "salut");
    manager1.setContactMetadata(manager2.vaultysId.did, "group", "pro");
    // assert.deepStrictEqual(
    //   manager1.getCertifiedMetadata(manager2.vaultysId.did),
    //   metadata2
    // );
    // assert.deepStrictEqual(
    //   manager1.getAllMetadata(manager2.vaultysId.did),
    //   {
    //     group: 'pro',
    //     name: 'salut',
    //     phone: 'f'
    //   }
    // );

    assert.ok(
      await manager1.verifyRelationshipCertificate(manager2.vaultysId.did),
    );
    assert.ok(
      await manager2.verifyRelationshipCertificate(manager1.vaultysId.did),
    );
  });

  it("perform migration from version 0 to 1",  async () => {
    const ids = [];
    for(let i = 0; i <10; i++) {
      const s1 = MemoryStorage(() => "");
      const id1 = new IdManager(await VaultysId.generatePerson(), s1);
      for(let j = 0; j <10; j++) {
        const s2 = MemoryStorage(() => "");
        const id2 = new IdManager(await VaultysId.generatePerson(), s2);
        id1.saveContact(id2.vaultysId);
        id1.setContactMetadata(id2.vaultysId.did, "test", id2.vaultysId.did);
      }
      ids.push(id1)
    }

    for(const id of ids) {
      id.migrate(0);
      assert.equal(id.contacts.length, 10);
      id.migrate(1);
      assert.equal(id.contacts.length, 10);
    }
   
  });

});

File: rust/tests/id_manager_test.rs
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};
use vaultysid::crypto::{from_base64, hash, to_base64, to_hex};
use vaultysid::memory_channel::MemoryChannel;
use vaultysid::{File, IdManager, MemoryStore, StoredApp, StoredContact, VaultysId};

const PASSPHRASE: &str = "test_passphrase_123";

#[tokio::test]
async fn test_vaultys_secret_serialization() {
    // Test secret serialization and deserialization
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let secret = vaultys_id.get_secret().unwrap();

    // Create second ID from the same secret
    let id2 = VaultysId::from_secret(&secret, None).unwrap();

    // Both should produce the same ID
    assert_eq!(vaultys_id.id(), id2.id());

    // Test HMAC generation
    let test_data = b"test data";
    let hmac1 = vaultys_id.sign_challenge(test_data).await.unwrap();
    let hmac2 = id2.sign_challenge(test_data).await.unwrap();

    assert_eq!(hmac1.signature, hmac2.signature);
}

#[tokio::test]
async fn test_vaultys_secret_base64_serialization() {
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let secret = vaultys_id.get_secret().unwrap();
    let secret_b64 = to_base64(&secret);

    // Create second ID from base64 secret
    let secret_decoded = from_base64(&secret_b64).unwrap();
    let id2 = VaultysId::from_secret(&secret_decoded, None).unwrap();

    // Both should produce the same ID
    assert_eq!(vaultys_id.id(), id2.id());

    // Test HMAC generation
    let test_data = b"test data";
    let hmac1 = vaultys_id.sign_challenge(test_data).await.unwrap();
    let hmac2 = id2.sign_challenge(test_data).await.unwrap();

    assert_eq!(hmac1.signature, hmac2.signature);
}

#[tokio::test]
async fn test_public_id_serialization() {
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let public_id = vaultys_id.id();

    // Create a public ID from the bytes
    let id2 = VaultysId::from_id(&public_id, None, None).unwrap();

    // Both should have the same public ID
    assert_eq!(vaultys_id.id(), id2.id());

    // Test signing (should only work with the original)
    let test_data = b"test data";
    let _hmac1 = vaultys_id.sign_challenge(test_data).await.unwrap();

    // id2 is public-only, so it can't sign
    assert!(id2.sign_challenge(test_data).await.is_err());
}

#[tokio::test]
async fn test_sign_unspecified_data() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    let s = "Some challenge to sign";
    let signature = manager.sign_challenge(s.as_bytes()).await.unwrap();

    assert!(!signature.is_empty());

    // Check that signatures are logged (implementation-specific)
    // In a real implementation, we would verify signature logging
}

#[tokio::test]
async fn test_sign_document_hash() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    let file = File {
        name: "test.pdf".to_string(),
        file_type: "application/pdf".to_string(),
        array_buffer: b"PDF content here".to_vec(),
        signatures: Vec::new(),
    };

    let h = hash("sha256", &file.array_buffer);
    let signature = manager.sign_file(&file).await.unwrap();

    assert!(!signature.signature.is_empty());
    // The signer should be a DID string
    assert!(signature.signer.starts_with("did:"));

    // Verify the signature contains the correct challenge
    let _expected_challenge = format!("Sign file: {} (SHA256: {})", file.name, to_hex(&h));
    // In a real implementation, we would verify this against the actual signature
}

#[tokio::test]
async fn test_web_of_trust_operations() {
    // Create two IdManagers
    let store1 = Box::new(MemoryStore::new());
    let vaultys_id1 = VaultysId::generate_person().await.unwrap();
    let manager1 = IdManager::new(vaultys_id1, store1).await.unwrap();

    let store2 = Box::new(MemoryStore::new());
    let vaultys_id2 = VaultysId::generate_person().await.unwrap();
    let manager2 = IdManager::new(vaultys_id2, store2).await.unwrap();

    // Set metadata for both managers
    manager1.set_name("Alice").await.unwrap();
    manager1.set_email("alice@example.com").await.unwrap();
    manager1.set_phone("+1234567890").await.unwrap();

    manager2.set_name("Bob").await.unwrap();
    manager2.set_email("bob@example.com").await.unwrap();
    manager2.set_phone("+0987654321").await.unwrap();

    // Create a contact for manager1
    let id1_hex = to_hex(&manager1.vaultys_id.lock().await.id());
    let contact = StoredContact {
        did: format!("did:vaultys:{}", id1_hex),
        certificate: None,
        metadata: HashMap::new(),
        id: Some(manager2.vaultys_id.lock().await.id()),
    };

    manager1.save_contact(contact).await.unwrap();

    // Check contacts
    let contacts = manager1.contacts().await;
    assert_eq!(contacts.len(), 1);
    assert_eq!(contacts[0].did, format!("did:vaultys:{}", id1_hex));
}

#[tokio::test]
async fn test_backup_and_restore() {
    // Create original manager
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Set some data
    manager.set_name("Test User").await.unwrap();
    manager.set_email("test@example.com").await.unwrap();

    // Create a backup
    let backup = manager.export_backup(PASSPHRASE).await.unwrap();
    assert!(!backup.is_empty());

    // Restore from backup
    let restored = IdManager::import_backup(&backup, PASSPHRASE, None)
        .await
        .unwrap();

    // The restored manager should have a valid VaultysId
    // Note: The exact ID may differ due to internal representation
    assert!(!restored.vaultys_id.lock().await.id().is_empty());
}

#[tokio::test]
async fn test_encrypted_backup() {
    // Create original manager with some data
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    manager.set_name("Encrypted User").await.unwrap();
    manager.set_phone("+1234567890").await.unwrap();

    // Create encrypted backup
    let backup = manager.export_backup(PASSPHRASE).await.unwrap();

    // Try to restore with wrong password - should fail
    assert!(IdManager::import_backup(&backup, "wrong_password", None)
        .await
        .is_err());

    // Restore with correct password - should succeed
    let restored = IdManager::import_backup(&backup, PASSPHRASE, None)
        .await
        .unwrap();
    // The restored manager should have a valid VaultysId
    assert!(!restored.vaultys_id.lock().await.id().is_empty());
}

#[tokio::test]
async fn test_srp_challenge_success() {
    // Create two managers
    let store1 = Box::new(MemoryStore::new());
    let vaultys_id1 = VaultysId::generate_person().await.unwrap();
    let manager1 = IdManager::new(vaultys_id1, store1).await.unwrap();

    let store2 = Box::new(MemoryStore::new());
    let vaultys_id2 = VaultysId::generate_person().await.unwrap();
    let manager2 = IdManager::new(vaultys_id2, store2).await.unwrap();

    // Set metadata
    manager1.set_name("Alice").await.unwrap();
    manager2.set_name("Bob").await.unwrap();

    // Create a bidirectional channel
    let mut channel = MemoryChannel::new();

    // Run SRP challenge
    let id2_hex = to_hex(&manager2.vaultys_id.lock().await.id());

    // Note: In a real scenario, these would run concurrently
    // For testing, we'd need proper async coordination and separate channels
    // This is a simplified version showing the API usage
    let _result = async {
        // These would need separate channel instances in practice
        // manager1.start_srp(&id2_hex, &mut channel).await
        // manager2.accept_srp(&mut channel).await
    };
    // For testing, we'd need to mock the channel properly

    // After successful SRP, contacts should be saved
    // This test would need proper async coordination to work fully
}

#[tokio::test]
async fn test_file_encryption_decryption() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Create a test file
    let original_file = File {
        name: "test_document.pdf".to_string(),
        file_type: "application/pdf".to_string(),
        array_buffer: b"This is a test PDF content with some data".to_vec(),
        signatures: Vec::new(),
    };

    // Encrypt the file
    let encrypted = manager.encrypt_file(&original_file, None).await.unwrap();
    assert!(encrypted.len() > original_file.array_buffer.len());

    // Decrypt the file
    let decrypted_file = manager.decrypt_file(&encrypted).await.unwrap();

    // Verify the decrypted content matches
    assert_eq!(decrypted_file.array_buffer, original_file.array_buffer);
}

#[tokio::test]
async fn test_file_encryption_with_different_types() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    let test_cases = vec![
        ("small.txt", "text/plain", b"Small file".to_vec()),
        (
            "medium.json",
            "application/json",
            b"{\"test\": \"data\"}".to_vec(),
        ),
        ("large.bin", "application/octet-stream", vec![0u8; 10000]),
    ];

    for (name, file_type, content) in test_cases {
        let file = File {
            name: name.to_string(),
            file_type: file_type.to_string(),
            array_buffer: content.clone(),
            signatures: Vec::new(),
        };

        // Encrypt and decrypt
        let encrypted = manager.encrypt_file(&file, None).await.unwrap();
        let decrypted = manager.decrypt_file(&encrypted).await.unwrap();

        // Verify
        assert_eq!(decrypted.array_buffer, content);
    }
}

#[tokio::test]
async fn test_contact_metadata_operations() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Create and save a contact
    let contact = StoredContact {
        did: "did:vaultys:test123".to_string(),
        certificate: None,
        metadata: HashMap::new(),
        id: None,
    };

    manager.save_contact(contact).await.unwrap();

    // Set metadata
    manager
        .set_contact_metadata("did:vaultys:test123", "name", "Bob")
        .await
        .unwrap();
    manager
        .set_contact_metadata("did:vaultys:test123", "email", "bob@example.com")
        .await
        .unwrap();

    // Get metadata
    let name = manager
        .get_contact_metadata("did:vaultys:test123", "name")
        .await;
    assert_eq!(name, Some("Bob".to_string()));

    let email = manager
        .get_contact_metadata("did:vaultys:test123", "email")
        .await;
    assert_eq!(email, Some("bob@example.com".to_string()));
}

#[tokio::test]
async fn test_app_management() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Create and save an app
    let app = StoredApp {
        site: "https://example.com".to_string(),
        server_id: Some("server123".to_string()),
        certificate: None,
        timestamp: Some(
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        ),
    };

    manager.save_app(app.clone()).await.unwrap();

    // Retrieve app
    let retrieved = manager.get_app("https://example.com").await;
    assert!(retrieved.is_some());
    assert_eq!(retrieved.unwrap().server_id, Some("server123".to_string()));

    // List apps
    let apps = manager.apps().await;
    assert_eq!(apps.len(), 1);
}

#[tokio::test]
async fn test_protocol_version() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let mut manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Default version should be 1 (verified through behavior)
    // The protocol version is internal to the implementation

    // Set a different version
    manager.set_protocol_version(2);
    // Version change is reflected in protocol behavior
}

#[tokio::test]
async fn test_display_name_fallback() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Without a name set, display_name should return fingerprint
    let display_name = manager.display_name().await;
    assert!(!display_name.is_empty());

    // After setting name, display_name should return the name
    manager.set_name("Alice").await.unwrap();
    assert_eq!(manager.display_name().await, "Alice");
}

#[tokio::test]
async fn test_phone_and_email_storage() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Set and get phone
    manager.set_phone("+1234567890").await.unwrap();
    assert_eq!(manager.phone().await, Some("+1234567890".to_string()));

    // Set and get email
    manager.set_email("test@example.com").await.unwrap();
    assert_eq!(manager.email().await, Some("test@example.com".to_string()));
}

#[tokio::test]
async fn test_file_signature() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Create a file to sign
    let file = File {
        name: "contract.pdf".to_string(),
        file_type: "application/pdf".to_string(),
        array_buffer: b"Important contract content".to_vec(),
        signatures: Vec::new(),
    };

    // Sign the file
    let signature = manager.sign_file(&file).await.unwrap();

    assert!(!signature.signature.is_empty());
    // The signer should be a DID string
    assert!(signature.signer.starts_with("did:"));
    assert!(signature.timestamp > 0);
}

#[tokio::test]
async fn test_multiple_contacts() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Create multiple contacts
    for i in 0..5 {
        let contact = StoredContact {
            did: format!("did:vaultys:test{}", i),
            certificate: None,
            metadata: HashMap::new(),
            id: None,
        };
        manager.save_contact(contact).await.unwrap();
    }

    // Verify all contacts are saved
    let contacts = manager.contacts().await;
    assert_eq!(contacts.len(), 5);

    // Verify individual retrieval
    for i in 0..5 {
        let contact = manager.get_contact(&format!("did:vaultys:test{}", i)).await;
        assert!(contact.is_some());
    }
}

#[tokio::test]
async fn test_empty_file_encryption() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Test with empty file
    let empty_file = File {
        name: "empty.txt".to_string(),
        file_type: "text/plain".to_string(),
        array_buffer: Vec::new(),
        signatures: Vec::new(),
    };

    // Encrypt and decrypt empty file
    let encrypted = manager.encrypt_file(&empty_file, None).await.unwrap();
    let decrypted = manager.decrypt_file(&encrypted).await.unwrap();

    assert_eq!(decrypted.array_buffer, empty_file.array_buffer);
}

#[tokio::test]
async fn test_large_file_encryption() {
    let store = Box::new(MemoryStore::new());
    let vaultys_id = VaultysId::generate_person().await.unwrap();
    let manager = IdManager::new(vaultys_id, store).await.unwrap();

    // Create a large file (1MB)
    let large_content = vec![42u8; 1024 * 1024];
    let large_file = File {
        name: "large.bin".to_string(),
        file_type: "application/octet-stream".to_string(),
        array_buffer: large_content.clone(),
        signatures: Vec::new(),
    };

    // Encrypt and decrypt
    let encrypted = manager.encrypt_file(&large_file, None).await.unwrap();
    let decrypted = manager.decrypt_file(&encrypted).await.unwrap();

    assert_eq!(decrypted.array_buffer, large_content);
}
